lib/pengine/unpack.c

```diff
@@ -2877,48 +2877,40 @@ static void
 unpack_migrate_to_success(pe_resource_t *rsc, const pe_node_t *node,
                           const xmlNode *xml_op, pe_working_set_t *data_set)
 {
-    /* A successful migration sequence is:
-     *    migrate_to on source node
-     *    migrate_from on target node
-     *    stop on source node
+    /* A complete migration sequence is:
+     * 1. migrate_to on source node (which succeeded if we get to this function)
+     * 2. migrate_from on target node
+     * 3. stop on source node
      *
-     * But there could be scenarios like (It's easier to produce with cluster
-     * property batch-limit=1):
-     *
-     * - rscA is live-migrating from node1 to node2.
-     *
-     * - Before migrate_to on node1 returns, put node2 into standby.
-     *
-     * - Transition aborts upon return of successful migrate_to on node1. New
-     *   transition is going to stop the rscA on both nodes and start it on
-     *   node1.
+     * If no migrate_from has happened, the migration is considered to be
+     * "partial". If the migrate_from succeeded but no stop has happened, the
+     * migration is considered to be "dangling".
      *
-     * - While it is stopping on node1, run something that is going to make
-     *   the transition abort again like:
-     *   crm_resource  --resource rscA --ban --node node2
+     * If a successful migrate_to and stop have happened on the source node, we
+     * still need to check for a partial migration, due to scenarios (easier to
+     * produce with batch-limit=1) like:
      *
-     * - Transition aborts upon return of stop on node1.
+     * - A resource is migrating from node1 to node2, and a migrate_to is
+     *   initiated for it on node1.
      *
-     * Now although there's a stop on node1, it's still a partial migration and
-     * rscA is still potentially active on node2.
+     * - node2 goes into standby mode while the migrate_to is pending, which
+     *   aborts the transition.
      *
-     * So even if a migrate_to is followed by a stop, we still need to check
-     * whether there's a corresponding migrate_from or any newer operation on
-     * the target.
+     * - Upon completion of the migrate_to, a new transition schedules a stop
+     *   on both nodes and a start on node1.
      *
-     * If no migrate_from has happened, the migration is considered to be
-     * "partial". If the migrate_from failed, make sure the resource gets
-     * stopped on both source and target (if up).
+     * - If the new transition is aborted for any reason while the resource is
+     *   stopping on node1, the transition after that stop completes will see
+     *   the migrate_from and stop on the source, but it's still a partial
```

Comment:
migrate_to and stop on the source?

Reply:
"on the source" refers only to the stop, I'll add "on the target" to migrate_from :)

Reply:
In this sequence, the `migrate_to` and the `stop` (source) have run, but the `migrate_from` never even began on the target. That's why I didn't resolve the ambiguity correctly. Is the `migrate_from` in the target's history (as pending or something), even though it was never initiated?

Reply:
Updated top commit to fix